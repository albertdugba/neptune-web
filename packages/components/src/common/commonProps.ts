export type CommonProps = {
  /**
   * Space-separated list of the case-sensitive CSS classes,
   * most would be applied to outermost element inside the component
   *
   * @see https://github.com/transferwise/neptune-web/blob/main/rfc/0001-always-pass-classname.md
   */
  className?: string;
};

export type AriaLabelProp = {
  'aria-label'?: string;
};

export type LinkProps = {
  href: string;
  target?: '_blank' | '_self' | '_parent' | '_top';
};
