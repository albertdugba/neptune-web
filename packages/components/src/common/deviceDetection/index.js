export { isIosDevice } from './deviceDetection';
export { isTouchDevice } from './isTouchDevice';
