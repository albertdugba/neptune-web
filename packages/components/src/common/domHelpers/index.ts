export function isServerSideRendering(): boolean {
  return typeof window === 'undefined';
}

export {
  addClickClassToDocumentOnIos,
  removeClickClassFromDocumentOnIos,
} from './documentIosClick';
