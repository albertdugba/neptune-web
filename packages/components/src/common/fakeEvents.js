export const fakeEvent = () => ({
  nativeEvent: { stopImmediatePropagation() {} },
  stopPropagation() {},
  preventDefault() {},
});

export const fakeKeyDownEventForKey = (keyCode) => ({
  ...fakeEvent(),
  keyCode,
});

export const fakeKeyDownEvent = (eventAttribute = {}) => ({
  ...fakeEvent(),
  ...eventAttribute,
});
