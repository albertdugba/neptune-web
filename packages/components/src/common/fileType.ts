export enum FileType {
  PDF = '.pdf,application/pdf',
  JPG = '.jpg,.jpeg,image/jpeg',
  PNG = '.png,image/png',
  CSV = '.csv,text/csv',
}
