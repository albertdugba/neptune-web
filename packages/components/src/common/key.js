export const Key = {
  ENTER: 'Enter',
  ESCAPE: 'Escape',
  SPACE: [' ', 'Spacebar'],
  TAB: 'Tab',
  PERIOD: '.',
  COMMA: ',',
};
