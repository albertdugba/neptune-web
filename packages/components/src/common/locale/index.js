export const DEFAULT_LANG = 'en';
export const DEFAULT_LOCALE = 'en-GB';

export const SUPPORTED_RTL = ['he-IL'];

export const SUPPORTED_LANGUAGES = [
  DEFAULT_LANG,
  'ja',
  'de',
  'es',
  'fr',
  'hu',
  'id',
  'it',
  'pl',
  'pt',
  'ro',
  'ru',
  'tr',
  'zh',
];

/**
 * Verifies and adjusts locale (replace `_` with `-`)
 * Returns null if locale is unrecognized by {Intl.Locale}
 *
 * @param locale (`es`, `es_ES`, `en-GB`, `en`, `ja`, `ja-JP` etc)
 * @returns {Intl.Locale.baseName|null}
 */
export function adjustLocale(locale) {
  if (!locale || locale.trim().length === 0) {
    return null;
  }
  try {
    const { baseName } = new Intl.Locale(locale.trim().replace('_', '-'));
    return baseName;
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error(error);
    return null;
  }
}

/**
 * Provides corresponding lang (iso2) for provided locale
 * if locale is invalid or language is unsupported returns null
 *
 * @param locale (`es`, `es-ES`, `en-GB`, `en`, `ja`, `ja-JP` etc)
 * @returns {Intl.Locale.language|null}
 */
export function getLangFromLocale(locale) {
  const adjustedLocale = adjustLocale(locale);
  if (adjustedLocale === null) {
    return null;
  }
  try {
    const { language } = new Intl.Locale(adjustedLocale);
    if (SUPPORTED_LANGUAGES.includes(language)) {
      return language;
    }
    return null;
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error(error);
    return null;
  }
}

/**
 * Provides corresponding country code (iso2) for locales code with explicit region value (`es-ES`, `en-GB`, `ja-JP` etc.)
 * if the value is invalid or missing region it returns null
 *
 * @param locale
 * @returns {string|null}
 */
export const getCountryFromLocale = (locale) => {
  const adjustedLocale = adjustLocale(locale);
  if (adjustedLocale === null) {
    return null;
  }
  try {
    const { region } = new Intl.Locale(adjustedLocale);
    return region ?? null;
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error(error);
    return null;
  }
};
