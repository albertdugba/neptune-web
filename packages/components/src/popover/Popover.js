import React, { useRef, useState, cloneElement } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { Position } from '../common';
import ResponsivePanel from '../common/responsivePanel';
import { logActionRequiredIf } from '../utilities';

import './Popover.css';

const Popover = ({ children, className, content, preferredPlacement, title }) => {
  logActionRequired({ preferredPlacement });
  const anchorRef = useRef(null);
  const [open, setOpen] = useState(false);

  const onClose = () => setOpen(false);

  return (
    <span className={classnames('np-popover', className)}>
      <span className="d-inline-block" ref={anchorRef}>
        {cloneElement(children, {
          onClick: () => {
            if (children?.props?.onClick) {
              children.props.onClick();
            }
            setOpen(!open);
          },
        })}
      </span>
      <ResponsivePanel
        open={open}
        anchorRef={anchorRef}
        position={deprecatedPlacements[preferredPlacement] || preferredPlacement}
        onClose={onClose}
        arrow
        className="np-popover__container"
      >
        <div className="np-popover__content" aria-hidden={!open} role="tooltip">
          {title && (
            <div role="heading" aria-level={1} className="np-popover__title m-b-1">
              {title}
            </div>
          )}
          {content}
        </div>
      </ResponsivePanel>
    </span>
  );
};

const logActionRequired = ({ preferredPlacement }) => {
  logActionRequiredIf(
    `Popover has deprecated ${preferredPlacement} value for the 'preferredPlacement' prop. Please use ${deprecatedPlacements[preferredPlacement]} instead.`,
    deprecatedPlacements[preferredPlacement],
  );
};

Popover.defaultProps = {
  className: undefined,
  preferredPlacement: Position.RIGHT,
  title: undefined,
};

Popover.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  content: PropTypes.node.isRequired,
  /**
   * `'left-top'` / `'right-top'` are deprecated use `Position.TOP` / `'top'` instead,
   * `'bottom-right'` / `'bottom-left'` are deprecated use `Position.BOTTOM` / `'bottom'` instead
   */
  preferredPlacement: PropTypes.oneOf([
    'top',
    'right',
    'bottom',
    'left',
    'left-top',
    'right-top',
    'bottom-right',
    'bottom-left',
  ]),
  title: PropTypes.node,
};

const deprecatedPlacements = {
  [Position.BOTTOM_LEFT]: Position.BOTTOM,
  [Position.BOTTOM_RIGHT]: Position.BOTTOM,
  [Position.LEFT_TOP]: Position.TOP,
  [Position.RIGHT_TOP]: Position.TOP,
};

export default Popover;
