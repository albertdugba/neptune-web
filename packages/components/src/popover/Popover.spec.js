import { Breakpoint, Position } from '../common';
import { render, screen, userEvent } from '../test-utils';

import Popover from './Popover';

jest.mock(
  'react-transition-group/CSSTransition',
  () => (props) => props.in ? props.children : null,
);

jest.mock(
  '../dimmer',
  () =>
    function ({ open, children }) {
      return open ? <div className="dimmer">{children}</div> : null;
    },
);

describe('Popover', () => {
  const props = {
    arrow: true,
    content: 'content',
    position: Position.BOTTOM,
    title: 'title',
  };

  let container;
  let rerender;

  describe('on desktop', () => {
    it('renders when is open', async () => {
      render(
        <Popover {...props}>
          <button type="button">Open</button>
        </Popover>,
      );

      userEvent.click(screen.getByText('Open'));

      await waitForPanel();
      expect(getPanel()).toMatchSnapshot();
    });

    it('renders title', async () => {
      ({ container, rerender } = render(
        <Popover {...props}>
          <button type="button">Open</button>
        </Popover>,
      ));

      userEvent.click(screen.getByText('Open'));
      await waitForPanel();

      expect(getTitle()).toBeInTheDocument();

      rerender(
        <Popover {...props} title={undefined}>
          <button type="button">Open</button>
        </Popover>,
      );

      userEvent.click(screen.getByText('Open'));

      expect(getTitle()).not.toBeInTheDocument();
    });

    it('renders Panel onClick', async () => {
      render(
        <Popover {...props}>
          <button type="button">Open</button>
        </Popover>,
      );

      expect(getPanel()).not.toBeInTheDocument();

      userEvent.click(screen.getByText('Open'));
      await waitForPanel();

      expect(getPanel()).toBeInTheDocument();
    });
  });

  describe('on mobile', () => {
    beforeEach(() => {
      global.innerWidth = Breakpoint.SMALL - 1;
    });

    it('renders when is open', () => {
      ({ container } = render(
        <Popover {...props}>
          <button type="button">Open</button>
        </Popover>,
      ));

      userEvent.click(screen.getByText('Open'));

      expect(container).toMatchSnapshot();
    });

    it('renders BottomSheet onClick', () => {
      render(
        <Popover {...props}>
          <button type="button">Open</button>
        </Popover>,
      );

      expect(getBottomSheet()).not.toBeInTheDocument();

      userEvent.click(screen.getByText('Open'));

      expect(getBottomSheet()).toBeInTheDocument();
    });
  });

  const waitForPanel = async () => screen.findByText('content');
  const getPanel = () => document.querySelector('.np-panel');
  const getBottomSheet = () => document.querySelector('.np-bottom-sheet');
  const getTitle = () => screen.queryByText('title');
});
