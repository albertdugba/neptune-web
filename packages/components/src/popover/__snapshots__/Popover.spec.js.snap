// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover on desktop renders when is open 1`] = `
<div
  class="np-panel np-panel--open np-popover__container"
  data-popper-escaped="true"
  data-popper-placement="right"
  data-popper-reference-hidden="true"
  style="position: absolute; left: 0px; top: 0px; transform: translate(16px, 0px);"
>
  <div
    class="np-panel__content"
  >
    <div
      aria-hidden="false"
      class="np-popover__content"
      role="tooltip"
    >
      <div
        aria-level="1"
        class="np-popover__title m-b-1"
        role="heading"
      >
        title
      </div>
      content
    </div>
    <div
      class="np-panel__arrow"
      style="position: absolute; top: 0px; transform: translate(0px, 0px);"
    />
  </div>
</div>
`;

exports[`Popover on mobile renders when is open 1`] = `
<div>
  <span
    class="np-popover"
  >
    <span
      class="d-inline-block"
    >
      <button
        type="button"
      >
        Open
      </button>
    </span>
    <div
      class="dimmer"
    >
      <div
        class="sliding-panel np-bottom-sheet np-popover__container"
      >
        <div
          class="np-bottom-sheet--top-bar"
        >
          <div
            class="np-bottom-sheet--handler"
          />
          <button
            aria-label="Close"
            class="np-close-button close btn-link text-no-decoration sr-only np-bottom-sheet--close-btn"
            type="button"
          >
            <span
              aria-hidden="true"
              class="tw-icon tw-icon-cross "
              role="presentation"
            >
              <svg
                fill="currentColor"
                focusable="false"
                height="16"
                width="16"
              >
                <path
                  d="M9.131 8l5.435-5.434-1.132-1.132L8 6.87 2.566 1.434 1.434 2.566 6.87 8l-5.435 5.434 1.132 1.132L8 9.13l5.434 5.435 1.132-1.132L9.13 8z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="np-bottom-sheet--content"
          style="max-height: calc(768px - 64px - 32px);"
        >
          <div
            aria-hidden="false"
            class="np-popover__content"
            role="tooltip"
          >
            <div
              aria-level="1"
              class="np-popover__title m-b-1"
              role="heading"
            >
              title
            </div>
            content
          </div>
        </div>
      </div>
    </div>
  </span>
</div>
`;
